src/repository.rs

```diff
@@ -117,136 +125,155 @@ impl<D> Repository<D> for FileSystemRepository<D>
 where
     D: DataInterchange,
 {
-    type TargetRead = File;
-
-    fn store_metadata<M>(
-        &self,
-        meta_path: &MetadataPath,
-        version: &MetadataVersion,
-        metadata: &SignedMetadata<D, M>,
-    ) -> Result<()>
+    fn store_metadata<'a, M>(
+        &'a self,
+        meta_path: &'a MetadataPath,
+        version: &'a MetadataVersion,
+        metadata: &'a SignedMetadata<D, M>,
+    ) -> TufFuture<'a, Result<()>>
     where
-        M: Metadata,
+        M: Metadata + 'static,
     {
-        Self::check::<M>(meta_path)?;
+        Box::pinned(
+            async move {
+                Self::check::<M>(meta_path)?;
 
-        let mut path = self.local_path.join("metadata");
-        path.extend(meta_path.components::<D>(version));
+                let mut path = self.local_path.join("metadata");
+                path.extend(meta_path.components::<D>(version));
 
-        if path.exists() {
-            debug!("Metadata path exists. Overwriting: {:?}", path);
-        }
+                if path.exists() {
+                    debug!("Metadata path exists. Overwriting: {:?}", path);
+                }
 
-        atomically_write(&path, |write| {
-            D::to_writer(write, metadata)?;
-            Ok(())
-        })
+                let mut temp_file = create_temp_file(&path)?;
+                D::to_writer(&mut temp_file, metadata)?;
+                temp_file.persist(&path)?;
+
+                Ok(())
+            },
+        )
     }
 
     /// Fetch signed metadata.
-    fn fetch_metadata<M>(
-        &self,
-        meta_path: &MetadataPath,
-        version: &MetadataVersion,
-        max_size: &Option<usize>,
+    fn fetch_metadata<'a, M>(
+        &'a self,
+        meta_path: &'a MetadataPath,
+        version: &'a MetadataVersion,
+        max_size: &'a Option<usize>,
         min_bytes_per_second: u32,
-        hash_data: Option<(&HashAlgorithm, HashValue)>,
-    ) -> Result<SignedMetadata<D, M>>
+        hash_data: Option<(&'static HashAlgorithm, HashValue)>,
+    ) -> TufFuture<'a, Result<SignedMetadata<D, M>>>
     where
-        M: Metadata,
+        M: Metadata + 'static,
     {
-        Self::check::<M>(meta_path)?;
+        Box::pinned(
+            async move {
+                Self::check::<M>(&meta_path)?;
 
-        let mut path = self.local_path.join("metadata");
-        path.extend(meta_path.components::<D>(&version));
+                let mut path = self.local_path.join("metadata");
+                path.extend(meta_path.components::<D>(&version));
 
-        let read = SafeReader::new(
-            File::open(&path)?,
-            max_size.unwrap_or(::std::usize::MAX) as u64,
-            min_bytes_per_second,
-            hash_data,
-        )?;
+                let mut reader = SafeReader::new(
+                    AllowStdIo::new(File::open(&path)?),
```

Comment:
This will block the event loop on doing file IO. I'd recommend using a dedicated thread for file IO.

Reply:
I don't have a suggestion, unfortunately-- it really is just a matter of having to spin up a separate thread and do the work there.